include/envoy/ssl/context_config.h

```diff
@@ -111,6 +112,16 @@ class ContextConfig {
    * @return The maximum TLS protocol version to negotiate.
    */
   virtual unsigned maxProtocolVersion() const PURE;
+
+  /**
+   * @return true of the config is valid.
+   */
+  virtual bool isValid() const PURE;
```

Comment:
why is this needed?

Reply:
This method is going to be used in next PR. Please refer to our high level design PR(https://github.com/envoyproxy/envoy/pull/3748/files). For example, ContextManagerImpl::createSslClientContext() needs this method to decide whether to create ssl client context. 
I find it is convenient to use this method in tests when I add tests into context_impl_test.cc.

Reply:
Sorry I don't have time to go through the other change. It seems a bit odd to have an `isValid()` on this. The context should always be valid or it should throw an exception, so I would change the logic and remove this. Is there something I'm missing here?

Reply:
Maybe isValid() is not accurate. It returns true if we have static secrets to create context, or we have created a dynamic secret provider and the provider has downloaded secrets already. We are going to check if secrets are ready or not before creating any context objects. If we have created a secret provider but that provider has not downloaded any secret, then we should not create context. Does that make sense? Is it better to rename it to isSecretAvailable()?

Reply:
We are considering that if the gRPC connection breaks when envoy is waiting for secrets, then maybe we need isValid() before creating context. I am going to remove isValid() from this PR, and come up with a better idea in the next PR.